crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3488,10 +3488,48 @@ impl<'db> TypeInferenceBuilder<'db> {
             ast::Expr::NumberLiteral(_literal) => Type::Todo,
             ast::Expr::BooleanLiteral(_literal) => Type::Todo,
 
-            // TODO: this may be a place we need to revisit with special forms.
             ast::Expr::Subscript(subscript) => {
-                self.infer_subscript_expression(subscript);
-                Type::Todo
+                let ast::ExprSubscript {
+                    value,
+                    slice,
+                    ctx: _,
+                    range: _,
+                } = subscript;
+
+                let value_ty = self.infer_expression(value);
+
+                if value_ty
+                    .into_class_literal_type()
+                    .is_some_and(|class| class.is_known(self.db, KnownClass::Tuple))
+                {
+                    // TODO:
+                    // - homogeneous tuples
+                    // - PEP 646
+                    match &**slice {
+                        ast::Expr::Tuple(tuple) => {
+                            let mut elements = vec![];
+                            let mut return_todo = false;
+                            for element in tuple {
+                                let element_ty = self.infer_type_expression(element);
+                                return_todo |= element_ty.is_todo();
```

Comment:
I thought we only use `Todo` if the inference itself is directly the consequence of something we aren't handling but not when we continue deriving other types from it... which seems to be the case here?

I think I would prefer a type `(str, Todo, int)` over just `Todo` because this gives me less information where the `Todo` is coming from and makes it harder to track down

Reply:
The issue here is that certain parameters can change the meaning of the entire tuple. If we only inferred Todo for a subelement of a tuple here, we'd interpret the annotation `tuple[str, ...]` as `tuple[str, Todo]` which means "a tuple of length two where the first element is a `str` and the second element is todo". But that's flat-out wrong — `tuple[str, ...]` actually means "a tuple of arbitrary length, in which all elements in the tuple are instances of `str`" — the `...` changes the meaning of the entire tuple (it becomes a homogenous tuple rather than a heterogeneous tuple, and we [don't support homogenous tuples yet](https://github.com/astral-sh/ruff/issues/13855)).

Hmmm... But maybe I can limit the situations where we return Todo for the entire tuple to certain special elements that we know change the meaning of the entire tuple... And just infer Todo for subelements rather than the whole tuple otherwise.

Reply:
I made this change. It didn't really make _much_ impact, because even `Unpack[tuple[int, ...]]` as a subelement can change the meaning of the entire tuple, so we still need to infer `Todo` for the entire tuple if an `Expr::Subscript` is found in the tuple that we infer to be `Todo`. But, there are now a few more tuples where we only infer `Todo` for a subelement rather than the whole tuple. And, I made the comments more expansive to explain why we need to do it this way :-)